src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,949 @@
+// Copyright 2020-2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::connectors::impls::gcs::api_client::{
+    ApiClient, DefaultApiClient, FileId, HttpTaskCommand, HttpTaskRequest,
+};
+use crate::connectors::impls::gcs::chunked_buffer::ChunkedBuffer;
+use crate::connectors::prelude::{
+    Attempt, ErrorKind, EventSerializer, Result, SinkAddr, SinkContext, SinkManagerBuilder,
+    SinkReply, Url,
+};
+use crate::connectors::sink::{AsyncSinkReply, ContraflowData, Sink};
+use crate::connectors::utils::url::HttpsDefaults;
+use crate::connectors::{
+    Alias, CodecReq, Connector, ConnectorBuilder, ConnectorConfig, ConnectorType, Context,
+};
+use crate::system::KillSwitch;
+use crate::{connectors, QSIZE};
+use async_std::channel::{bounded, Receiver, Sender};
+use http_client::h1::H1Client;
+use http_client::HttpClient;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+use tremor_common::time::nanotime;
+use tremor_pipeline::{ConfigImpl, Event};
+use tremor_value::Value;
+use value_trait::ValueAccess;
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Config {
+    #[serde(default = "default_endpoint")]
+    endpoint: Url<HttpsDefaults>,
+    #[serde(default = "default_connect_timeout")]
+    connect_timeout: u64,
+    #[serde(default = "default_buffer_size")]
+    buffer_size: usize,
+    bucket: Option<String>,
+}
+
+#[allow(clippy::unwrap_used)]
+fn default_endpoint() -> Url<HttpsDefaults> {
```

Comment:
Would be nice to add a test-case for this to guarantee if ever an error sneaks in here we find it during testing